### PR TITLE
log, metrics: refactor metrics and log system for dm and other repositories to use

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
1. Now dumpling only has only one global logger. DM may init several dumpling loggers, which may cause a data race in dumpling. Besides, each logger has a different task name, which may be confusing if we only have one global logger.
2. DM can't set customized `conf.Labels` because we have preset their labels in `metrics.go`.

### What is changed and how it works?
1. Use a customized logger for each dumper instance.
2. Don't set prometheus vector at first.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### Release note

- No release note

### CODE REVIEW

File: v4/context/context.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"context"
+	"time"
+
+	"github.com/pingcap/dumpling/v4/log"
+)
+
+// Context is used to in dm to record some context field like
+// * go context
+// * logger
+type Context struct {
```
Comment (lance6716) [inline v4/context/context.go:16]: how about

```
type Context struct {
	context.Context
	logger log.Logger
}
```

so we don't need `(*Context).Context()`. and logger could only be accessed by `L()`

File: v4/export/metrics.go [hunk 2]
```diff
@@ -10,61 +10,80 @@ import (
...
+	finishedRowsCounter            *prometheus.CounterVec
+	finishedTablesCounter          *prometheus.CounterVec
+	writeTimeHistogram             *prometheus.HistogramVec
+	receiveWriteChunkTimeHistogram *prometheus.HistogramVec
+	errorCount                     *prometheus.CounterVec
+	taskChannelCapacity            *prometheus.GaugeVec
+)
+
+// InitMetricsVector inits metrics vectors.
+// This function much run before RegisterMetrics
```
Comment (lance6716) [inline v4/export/metrics.go:23]:
```suggestion
// This function must run before RegisterMetrics
```

File: v4/export/metrics.go [hunk 3]
```diff
@@ -10,61 +10,80 @@ import (
...
+func InitMetricsVector(labels prometheus.Labels) {
```
Comment (lance6716) [inline v4/export/metrics.go:24]: forget why we need change this 😂

File: v4/export/config.go [hunk 4]
```diff
@@ -172,8 +172,8 @@ func DefaultConfig() *Config {
 // String returns dumpling's config in json format
 func (conf *Config) String() string {
 	cfg, err := json.Marshal(conf)
-	if err != nil {
-		log.Error("fail to marshal config to json", zap.Error(err))
+	if err != nil && conf.Logger != nil {
```
Comment (lance6716) [inline v4/export/config.go:175]: didn't find where `conf.Logger` is set

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: v4/export/config.go (hunk 4 above)
Comment (lichunzhu) [reply v4/export/config.go:175]: `conf.Logger` is set in DM, outside dumpling repository.

File: v4/context/context.go (hunk 1 above)
Comment (lichunzhu) [reply v4/context/context.go:15]: addressed in https://github.com/pingcap/dumpling/pull/242/commits/fcc2ab6e9f673aa3a09baea0d4cae588cd9fbaaa

**Comment by lance6716** [comment]
/lgtm

need remove some call of `Context()` in test files

**Review by lance6716 — APPROVED** [review]

**Pull request merged**